File: pyzk/ir/ir_pass/dead_code_elimination.py
from pyzk.ir.ir_builder import IRGraph
from pyzk.ir.ir_pass.abstract_pass import AbstractIRPass


class DeadCodeEliminationIRPass(AbstractIRPass):
    def __init__(self):
        super().__init__()

    def exec(self, ir_graph: IRGraph) -> IRGraph:
        stmts = ir_graph.export_stmts()
        ensuring_keep_stmts = [stmt.operator.dce_keep() for stmt in stmts]
        in_d, out_d = ir_graph.get_io_degrees()
        killing_queue = []
        to_be_eliminated = []
        for i, stmt in enumerate(stmts):
            if out_d[i] == 0 and not ensuring_keep_stmts[i]:
                killing_queue.append(i)
        while len(killing_queue) > 0:
            to_be_killed = killing_queue.pop()
            to_be_eliminated.append(to_be_killed)
            referring_tos = stmts[to_be_killed].arguments
            for k, t in referring_tos.items():
                if t is None:
                    continue
                in_d[to_be_killed] -= 1
                out_d[t] -= 1
                if out_d[t] == 0 and not ensuring_keep_stmts[t]:
                    killing_queue.append(t)
        ir_graph.remove_stmt_bunch(to_be_eliminated)
        return ir_graph

File: src/optim/dead_code_elimination.rs
use crate::ir::IRGraph;
use crate::types::StmtId;

use super::IRPass;

pub struct DeadCodeElimination;

impl IRPass for DeadCodeElimination {
    fn exec(&self, mut ir_graph: IRGraph) -> IRGraph {
        let n = ir_graph.stmts.len();
        let ensure_keep: Vec<bool> = ir_graph.stmts.iter().map(|s| s.ir.is_fixed()).collect();
        let (mut _in_d, mut out_d) = ir_graph.get_io_degrees();

        let mut killing_queue: Vec<usize> = Vec::new();
        let mut to_eliminate: Vec<StmtId> = Vec::new();

        for i in 0..n {
            if out_d[i] == 0 && !ensure_keep[i] {
                killing_queue.push(i);
            }
        }

        while let Some(idx) = killing_queue.pop() {
            to_eliminate.push(idx as StmtId);
            for &arg in &ir_graph.stmts[idx].arguments {
                out_d[arg as usize] -= 1;
                if out_d[arg as usize] == 0 && !ensure_keep[arg as usize] {
                    killing_queue.push(arg as usize);
                }
            }
        }

        ir_graph.remove_stmt_bunch(&to_eliminate);
        ir_graph
    }
}
